Think the failing test is related; https://jenkins.dockerproject.org/job/Docker-PRs/24958/console


```
10:01:44 ?   	github.com/docker/docker/contrib/httpserver	[no test files]
10:02:37 --- FAIL: TestValidateConfiguration (0.00s)
10:02:37 	config_test.go:243: valid DNS, got validation error
10:02:37 time="2016-03-11T10:02:37Z" level=warning msg="Changing requested CPUShares of 1 to minimum allowed of 2" 
10:02:37 time="2016-03-11T10:02:37Z" level=warning msg="Changing requested CPUShares of 262145 to maximum allowed of 262144" 
10:02:37 FAIL
10:02:37 coverage: 7.8% of statements
10:02:37 FAIL	github.com/docker/docker/daemon	0.074s
```

--- a/daemon/config.go
+++ b/daemon/config.go
@@ -164,6 +164,11 @@ func ReloadConfiguration(configFile string, flags *flag.FlagSet, reload func(*Co
 	if err != nil {
 		return err
 	}
+
+	if err := validateConfiguration(newConfig); err != nil {
+		return fmt.Errorf("file configuration validation failed (%v)", err)
+	}
+
 	reload(newConfig)
 	return nil
 }
@@ -182,6 +187,10 @@ func MergeDaemonConfigurations(flagsConfig *Config, flags *flag.FlagSet, configF
 	fileConfig, err := getConflictFreeConfiguration(configFile, flags)
 	if err != nil {
 		return nil, err
+	}
+
+	if err := validateConfiguration(fileConfig); err != nil {
+		return nil, fmt.Errorf("file configuration validation failed (%v)", err)
 	}
 
 	// merge flags configuration on top of the file configuration
@@ -334,5 +343,32 @@ func findConfigurationConflicts(config map[string]interface{}, flags *flag.FlagS
 	if len(conflicts) > 0 {
 		return fmt.Errorf("the following directives are specified both as a flag and in the configuration file: %s", strings.Join(conflicts, ", "))
 	}
+	return nil
+}
+
+// validateConfiguration validates some specific configs.
+// such as config.DNS, config.Labels, config.DNSSearch
+func validateConfiguration(config *Config) error {
+	// validate DNS
+	for _, dns := range config.DNS {
+		if _, err := opts.ValidateIPAddress(dns); err != nil {
+			return err
+		}
+	}
+
+	// validate DNSSearch
+	for _, dnsSearch := range config.DNSSearch {
+		if _, err := opts.ValidateDNSSearch(dnsSearch); err != nil {
+			return err
+		}
+	}
+
+	// validate Labels
+	for _, label := range config.Labels {
+		if _, err := opts.ValidateLabel(label); err != nil {
+			return err
+		}
+	}
+
 	return nil
 }

--- a/daemon/config_test.go
+++ b/daemon/config_test.go
@@ -208,3 +208,71 @@ func TestFindConfigurationConflictsWithMergedValues(t *testing.T) {
 		t.Fatalf("expected hosts conflict, got %v", err)
 	}
 }
+
+func TestValidateConfiguration(t *testing.T) {
+	c1 := &Config{
+		CommonConfig: CommonConfig{
+			Labels: []string{"one"},
+		},
+	}
+
+	err := validateConfiguration(c1)
+	if err == nil {
+		t.Fatal("expected error, got nil")
+	}
+
+	c2 := &Config{
+		CommonConfig: CommonConfig{
+			Labels: []string{"one=two"},
+		},
+	}
+
+	err = validateConfiguration(c2)
+	if err != nil {
+		t.Fatalf("expected no error, got error %v", err)
+	}
+
+	c3 := &Config{
+		CommonConfig: CommonConfig{
+			DNS: []string{"1.1.1.1"},
+		},
+	}
+
+	err = validateConfiguration(c3)
+	if err != nil {
+		t.Fatalf("expected no error, got error %v", err)
+	}
+
+	c4 := &Config{
+		CommonConfig: CommonConfig{
+			DNS: []string{"1.1.1.1o"},
+		},
+	}
+
+	err = validateConfiguration(c4)
+	if err == nil {
+		t.Fatal("expected error, got nil")
+	}
+
+	c5 := &Config{
+		CommonConfig: CommonConfig{
+			DNSSearch: []string{"a.b.c"},
+		},
+	}
+
+	err = validateConfiguration(c5)
+	if err != nil {
+		t.Fatalf("expected no error, got error %v", err)
+	}
+
+	c6 := &Config{
+		CommonConfig: CommonConfig{
+			DNSSearch: []string{"123456"},
+		},
+	}
+
+	err = validateConfiguration(c6)
+	if err == nil {
+		t.Fatal("expected error, got nil")
+	}
+}

